File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
/*
    목표위치로 갈 때 최소시간을 구하는 4방향 BFS
*/

import java.util.*;

class Solution {
    public int solution(int[][] maps) {
        int dx [] = {-1,1,0,0};
        int dy [] = {0,0,-1,1};
        int time=-1;
        boolean visited [][] = new boolean[maps.length][maps[0].length];
        
        PriorityQueue<Robot> pq = new PriorityQueue<>();
        pq.add(new Robot(0,0,0));
        
        while(!pq.isEmpty()){
            Robot curr = pq.poll();
            
            //목표 위치에 다다른 경우
            if(curr.row == maps.length-1 && curr.col == maps[0].length-1){
                time = curr.time+1;
                break;
            }
            
            for(int k=0;k<dx.length;k++){
                Robot next = new Robot(curr.row + dx[k], curr.col + dy[k], curr.time+1);
               
                //배열 밖으로 벗어나는 경우
                if(next.row<0 || next.col<0 || next.row>=maps.length || next.col>=maps[0].length){
                    continue;
                }
                //벽인 경우
                if(maps[next.row][next.col]==0){
  
                    continue;
                }
                //이미 방문한 경우
                if(visited[next.row][next.col]){ 
                    continue;
                }
                visited[next.row][next.col] = true;
                pq.add(next);
            }
        }
            
        return time;
    }
}
class Robot implements Comparable<Robot>{
    int row;
    int col;
    int time;
    
    public Robot(int row,int col, int time){
        this.row=row;
        this.col=col;
        this.time=time;
    }
    
    @Override
    public int compareTo(Robot o){
        return this.time - o.time;
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
import queue

class Coordinate:
    def __init__(self, row, col,count):
        self.row = row
        self.col = col
        self.count = count
        
def solution(maps):
    q = queue.Queue()
    
    q.put(Coordinate(0,0,1))
    
    nextDirection = [[-1,0],[1,0],[0,-1],[0,1]]
    visited = [[False for col in range(len(maps[0]))] for row in range(len(maps))]
    
    answer = -1
    
    while q.qsize() > 0 :
        curr = q.get()
        visited[curr.row][curr.col] = True
        
        if curr.row == len(maps) -1 and curr.col == len(maps[0]) - 1:
            answer = curr.count
            break
        
        for n in nextDirection:
            next = Coordinate(curr.row + n[0], curr.col+n[1],curr.count + 1)
            if next.row < 0 or next.row >= len(maps) or next.col < 0 or next.col >= len(maps[0]):
                continue
            if maps[next.row][next.col] == 0:
                continue
            if visited[next.row][next.col] == True:
                continue
            visited[next.row][next.col] = True
            q.put(next)
    
    return answer